src/Controller/ResetPasswordController.php

```diff
@@ -0,0 +1,177 @@
+use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
+use Symfony\Component\Routing\Attribute\Route;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+use function Symfony\Component\Clock\now;
+
+class ResetPasswordController extends AbstractController
+{
+    private const EXPIRES_AFTER = 3600; // 1 hour
+    private const SESSION_ID = 'password_reset_id';
+    private const SUBHASH_PARAMETER = 'sub-hash';
+
+    public function __construct(
+        #[Lazy]
+        private UriSigner $uriSigner,
+
+        #[Lazy]
+        private MailerInterface $mailer,
+
+        #[Lazy]
+        private UserRepository $users,
+
+        #[Lazy]
+        private UserPasswordHasherInterface $passwordHasher,
+
+        #[Lazy]
+        private EntityManagerInterface $em,
+
+        #[Lazy]
+        private Security $security,
+    ) {
+    }
+
+    #[Route('/forgot-password', name: 'password_reset_request')]
+    public function requestReset(
+        Request $request,
+        Session $session,
+    ): Response {
+        $form = $this->createForm(ForgotPasswordForm::class);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            /** @var string $email */
+            $email = $form->get('email')->getData();
+
+            $this->sendResetLink($email);
+
+            $session->getFlashBag()->add('note', \sprintf('A password reset link has been sent to %s.', $email));
+
+            return $this->redirectToRoute('homepage');
+        }
+
+        return $this->render('security/forgot_password.html.twig', [
+            'form' => $form,
+        ]);
+    }
+
+    #[Route('/reset-password/{id<\d+>}', name: 'password_reset')]
+    public function reset(
+        Request $request,
+        Session $session,
+        ?int $id = null,
+    ): Response {
+        if (null !== $id) {
+            return $this->addIdToSessionAndRedirect($request, $session, $id);
+        }
+
+        if (!$user = $this->users->find($session->get(self::SESSION_ID, 0))) {
+            throw $this->createNotFoundException('User not found');
+        }
+
+        if (!hash_equals($this->computeAlreadyUsedHash($user), $request->query->getString(self::SUBHASH_PARAMETER))) {
+            $session->getFlashBag()->add('error', 'This password reset link has already been used.');
+
+            return $this->redirectToRoute('password_reset_request');
+        }
+
+        $form = $this->createForm(ChangePasswordForm::class);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            /** @var string $plainPassword */
+            $plainPassword = $form->get('plainPassword')->getData();
+
+            $user->setPassword($this->passwordHasher->hashPassword($user, $plainPassword));
+            $this->em->flush();
+            $session->remove(self::SESSION_ID);
+            $this->security->login($user, 'form_login');
```

Comment:
Having to pass the `form_login` authenticator name here is a bummer... If you don't pass, it doesn't know what authenticator to use: `form_login` or `remember_me`.

Would be nice to configure a *default* authenticator for programmatic login.

Reply:
FYI, this has been fixed in 6.4+: the `remember_me` authenticator is now ignored. Removed these calls in ba7635f349861b15c120d2107d1103f2011b64ab